src/broker.rs

```diff
@@ -0,0 +1,91 @@
+use crate::{warn, MQTT_INSECURE_DEFAULT_PORT};
+use embedded_nal::{nb, AddrType, Dns, IpAddr, Ipv4Addr, SocketAddr};
+
+/// A type that allows us to (eventually) determine the broker address.
+pub trait Broker {
+    /// Retrieve the broker address (if available).
+    fn get_address(&mut self) -> Option<SocketAddr>;
+
+    /// Set the port of the broker connection.
+    fn set_port(&mut self, port: u16);
+}
+
+/// A broker that is specified using a qualified domain-name. The name will be resolved at some
+/// point in the future.
+pub struct NamedBroker<R: Dns, const T: usize = 255> {
+    raw: heapless::String<T>,
+    resolver: R,
+    addr: SocketAddr,
+}
+
+impl<R: Dns, const T: usize> NamedBroker<R, T> {
+    /// Construct a new named broker.
+    ///
+    /// # Args
+    /// * `broker` - The domain name of the broker, such as `broker.example.com`
+    /// * `resolver` - A [embedded_nal::Dns] resolver to resolve the broker domain name to an IP
+    /// address.
+    pub fn new(broker: &str, resolver: R) -> Self {
+        let addr: Ipv4Addr = broker.parse().unwrap_or(Ipv4Addr::UNSPECIFIED);
+
+        Self {
+            raw: heapless::String::from(broker),
+            resolver,
+            addr: SocketAddr::new(IpAddr::V4(addr), MQTT_INSECURE_DEFAULT_PORT),
+        }
+    }
+}
+
+impl<R: Dns, const T: usize> Broker for NamedBroker<R, T> {
+    fn get_address(&mut self) -> Option<SocketAddr> {
+        // Attempt to resolve the address.
+        if self.addr.ip().is_unspecified() {
+            match self.resolver.get_host_by_name(&self.raw, AddrType::IPv4) {
+                Ok(ip) => self.addr.set_ip(ip),
+                Err(nb::Error::WouldBlock) => {}
+                Err(_other) => {
+                    warn!("DNS lookup failed: {_other:?}")
+                }
+            }
+        }
+
+        if !self.addr.ip().is_unspecified() {
+            Some(self.addr)
+        } else {
+            None
+        }
+    }
+
+    fn set_port(&mut self, port: u16) {
+        self.addr.set_port(port)
+    }
+}
+
```

Comment:
Ah. Because of the resolver thingy. Might really be smoother if that was part of the stack.

Reply:
Ive thought about that, but I don't see how you'd support stacks that don't support DNS in that case.

Reply:
They'd be able to error, no?

Reply:
So you're saying that the `tcp_connect()` function would accept a string instead of a socket address? Or some type of `enum` specifying either a DNS name/port or IP/port? Interesting idea.

Reply:
Yes. Like other stacks.
https://doc.rust-lang.org/stable/std/net/struct.TcpStream.html#method.connect